### ISSUE TITLE
Voronoi Diagrams/Delaunay Triangulations

### ISSUE THREAD

**Issue by gampleman** [ianmackenzie/elm-geometry#38]
This proposes adding support for computing Voronoi Diagrams or Delaunay Triangulations or both.

Voronoi diagrams have many usecases, mine relates to uses in data visualization - as I would like to expose an interface aimed at those usecases in [elm-visualization](/gampleman/elm-visualization).

An important usecase there is to be able to quickly compute the closest point to another point (optionally limited by a radius), for example to [support mouse interactions with a data set with many points](https://bl.ocks.org/mbostock/8033015). However it can be used also to [help compute labels for a scatterplot](https://bl.ocks.org/mbostock/6909318).

These usecases impact a design of an API:

1. The resulting Voronoi should be able to serve as an index giving a fast way to find the nearest site to a point.
2. The api should be fast. AFAIK the runtime complexity should be O(nlogn).
3. It would help if the API allows incremental construction or even modification rather than rebuilding the Voronoi from scratch every time the data changes.

From my brief research on this topic, I found [this algorithm](http://www.personal.kent.edu/%7Ermuhamma/Compgeometry/MyCG/Voronoi/DivConqVor/divConqVor.htm) that should satisfy 2 and 3 above.

This algorithm needs to compute the [convex hull](http://www.personal.kent.edu/%7Ermuhamma/Compgeometry/MyCG/ConvexHull/convexHull.htm), which seems like that would be a useful addition in its own right.

----

Some other things that might be worth considering:

1. Supporting different distance metrics. An obvious use case seems [geographical voronoi](https://www.jasondavies.com/maps/voronoi/airports/) and weighted euclidian - i.e. for taking an importance metric into consideration (I believe this is useful for Voronoi treemaps).
2. Supporting 3 or more dimensions. I personally have no usecase for this (apart from the weighted usecase which can be expressed as a 3 dimensional Voronoi projected onto a plane).

----

Resources:

- [A textbook chapter on Voronoi diagrams](http://www.cs.uu.nl/geobook/voronoi.ps)
- [A textbook chapter on Delaunay Triangulations](http://www.cs.uu.nl/geobook/interpolation.pdf)

**Comment by ianmackenzie** [comment]
Probably won't get to this right away, but I had some ideas for a proposed API after the discussion on Slack yesterday and wanted to write them down before I forgot:
```elm
-- Plain Delaunay triangulation
Point2d.delaunayTriangulation : List Point2d -> TriangularMesh Point2d

-- Delaunay triangulation of a list of *vertices* which may have additional data
-- associated with them (IDs, colors etc.)
Point2d.delaunayTriangulationBy : (a -> Point2d) -> List a -> TriangularMesh a
```

**Comment by ianmackenzie** [comment]
Also, just looking at your comments again, I think an [R-tree](https://en.wikipedia.org/wiki/R-tree) or even a [K-d tree](https://en.wikipedia.org/wiki/K-d_tree) would be a simpler way to support 'nearest point' queries. Those are data structures that are very generally useful and so are likely to be added to `elm-geometry` (or in their own package) before Delaunay triangulation or Voronoi diagram generation. And they should definitely support 'nearest point' calculations in O(log n) time, with O(n log n) construction time.

**Comment by ianmackenzie** [comment]
If the Bowyer-Watson algorithm is used for Delaunay triangulation, it's necessary to be careful to ensure the outermost vertices are triangulated properly (it's easy to accidentally end up with some concave sections in the triangulation). There's what seems to be a nice solution at https://stackoverflow.com/a/36992359.

**Comment by ianmackenzie** [comment]
Implemented in #58.
